src/System.Management.Automation/engine/InternalCommands.cs

```diff
@@ -424,24 +431,59 @@ private void InitParallelParameterSet()
                 _taskJob = new PSTaskJob(
                     Parallel.ToString(),
                     ThrottleLimit);
+
+                return;
             }
-            else
+
+            // Set up for synchronous processing and data streaming.
+            _taskCollection = new PSDataCollection<System.Management.Automation.PSTasks.PSTask>();
+            _taskDataStreamWriter = new PSTaskDataStreamWriter(this);
+            _taskPool = new PSTaskPool(ThrottleLimit);
+            _taskPool.PoolComplete += (sender, args) =>
             {
-                _taskDataStreamWriter = new PSTaskDataStreamWriter(this);
-                _taskPool = new PSTaskPool(ThrottleLimit);
-                _taskPool.PoolComplete += (sender, args) =>
-                {
-                    _taskDataStreamWriter.Close();
-                };
-                if (TimeoutSeconds != 0)
-                {
-                    _taskTimer = new Timer(
-                        (_) => _taskPool.StopAll(),
-                        null,
-                        TimeoutSeconds * 1000,
-                        Timeout.Infinite);
-                }
+                _taskDataStreamWriter.Close();
+            };
+
+            // Create timeout timer if requested.
+            if (TimeoutSeconds != 0)
+            {
+                _taskTimer = new Timer(
+                    callback: (_) => { _taskCollection.Complete(); _taskPool.StopAll(); },
+                    state: null,
+                    dueTime: TimeoutSeconds * 1000,
+                    period: Timeout.Infinite);
             }
+
+            // Task collection handler.
+            System.Threading.ThreadPool.QueueUserWorkItem(
```

Comment:
I feel that the behaviour here, being threaded and slightly subtle (since there's a `while (true)`/`WaitHandle.WaitOne()`/`try`/`catch`) could use some comments both above the `QueueWorkItem` describing what the work item is and within the delegate describing

- Why `while (true)`
- Why `WaitOne()`
- What the `foreach` is doing
- What we're hoping to catch and why we catch everything (what happens if it's another exception?)

Reply:
I have a try/catch to ensure the task collection and task pool close gracefully on an unexpected error and allow running tasks to complete and get cleaned up.

I can't decide if an error should be written here.

But I will add comments.